File: apps/desktop/src/registries/BrandLogoRegistry.ts
import { ASSET_MANIFEST, BUNDLED_BRAND_LOGO_KEYS, type BundledBrandLogoKey } from './AssetManifest';
import { resolvePublicAsset } from '../utils/resolvePublicAsset';

const KNOWN_BRAND_KEYS = new Set<BundledBrandLogoKey>(BUNDLED_BRAND_LOGO_KEYS);

export class BrandLogoRegistry {
  /** All bundled brand logo keys shipped with the desktop app (excludes default). */
  static listBundledBrandKeys(): BundledBrandLogoKey[] {
    return [...BUNDLED_BRAND_LOGO_KEYS];
  }

  /** Normalize catalogue / inventory brand names to bundled asset keys. */
  static normalizeBrandKey(brandName: string): BundledBrandLogoKey | 'default' {
    const trimmed = brandName.trim().toLowerCase();
    if (KNOWN_BRAND_KEYS.has(trimmed as BundledBrandLogoKey)) {
      return trimmed as BundledBrandLogoKey;
    }

    const firstToken = trimmed.split(/[\s\-_/]+/).find((part) => part.length > 0);
    if (firstToken && KNOWN_BRAND_KEYS.has(firstToken as BundledBrandLogoKey)) {
      return firstToken as BundledBrandLogoKey;
    }

    for (const key of KNOWN_BRAND_KEYS) {
      if (trimmed.includes(key)) {
        return key;
      }
    }

    return 'default';
  }

  /** Suggest a bundled logo filename from a brand display name (e.g. "Asus" → "asus.svg"). */
  static logoFilenameForBrand(brandName: string): string | null {
    const normalized = BrandLogoRegistry.normalizeBrandKey(brandName);
    if (normalized !== 'default') {
      return `${normalized}.svg`;
    }
    return null;
  }

  /**
   * True when logo_filename points at a server-managed, user-uploaded logo
   * (written by the backend as `/assets/brand-logos/brand-{id}.{ext}`) rather
   * than a bundled asset. Uploaded logos are fetched through the authenticated
   * image proxy; bundled ones resolve to local assets.
   */
  static isUploadedLogo(logoFilename?: string | null): boolean {
    const trimmed = logoFilename?.trim();
    return Boolean(trimmed && trimmed.startsWith('/assets/brand-logos/brand-'));
  }

  /** Prefer explicit logo_filename; otherwise match bundled assets by brand name. */
  static resolveLogoFilename(brandName: string, logoFilename?: string | null): string | null {
    const trimmed = logoFilename?.trim();
    if (trimmed) {
      return trimmed.split('/').pop() ?? trimmed;
    }
    return BrandLogoRegistry.logoFilenameForBrand(brandName);
  }

  static getLogo(brandName: string): string {
    const normalized = BrandLogoRegistry.normalizeBrandKey(brandName);
    if (normalized !== 'default') {
      return resolvePublicAsset(ASSET_MANIFEST.brandLogos[normalized]);
    }
    return resolvePublicAsset(ASSET_MANIFEST.brandLogos.default);
  }

  static getFallbackLogo(brandName: string): string {
    const normalized = BrandLogoRegistry.normalizeBrandKey(brandName);
    if (normalized !== 'default' && normalized in ASSET_MANIFEST.brandLogosPng) {
      return resolvePublicAsset(ASSET_MANIFEST.brandLogosPng[normalized]);
    }
    return resolvePublicAsset(ASSET_MANIFEST.brandLogosPng.default);
  }
}

File: apps/mobile_flutter/lib/shared/catalogue/brand_logo_registry.dart
/// Brand logo asset paths — mirrors desktop `BrandLogoRegistry` / `AssetManifest`.
///
/// Note: bundled `*.png` files in this repo are SVG content (desktop convention).
/// Mobile always loads `.svg` via [flutter_svg] for correct rendering on Android/iOS.
abstract final class BrandLogoRegistry {
  static const _base = 'assets/brand-logos';

  static const Set<String> _knownKeys = {
    'asus',
    'acer',
    'hp',
    'dell',
    'lenovo',
    'msi',
    'apple',
    'sandisk',
    'logitech',
    'tp-link',
    'canon',
    'epson',
    'brother',
  };

  static String normalizeBrandKey(String brandName) {
    final trimmed = brandName.trim().toLowerCase();
    if (_knownKeys.contains(trimmed)) return trimmed;

    final firstToken = trimmed.split(RegExp(r'[\s\-_/]+')).firstWhere((part) => part.isNotEmpty, orElse: () => trimmed);
    if (_knownKeys.contains(firstToken)) return firstToken;

    for (final key in _knownKeys) {
      if (trimmed.contains(key)) return key;
    }
    return firstToken;
  }

  static String svgForName(String brandName) {
    final key = normalizeBrandKey(brandName);
    if (_knownKeys.contains(key)) return '$_base/$key.svg';
    return '$_base/default.svg';
  }

  static String assetForFilename(String? logoFilename) {
    if (logoFilename == null || logoFilename.trim().isEmpty) {
      return '$_base/default.svg';
    }
    final file = logoFilename.trim().split('/').last;
    if (file.startsWith('assets/')) return file;
    // Catalogue may store `.png` names that are SVG bytes — prefer `.svg` sibling.
    if (file.toLowerCase().endsWith('.png')) {
      final stem = file.substring(0, file.length - 4);
      if (_knownKeys.contains(stem.toLowerCase())) {
        return '$_base/$stem.svg';
      }
    }
    return '$_base/$file';
  }

  static List<String> assetCandidates({required String brandName, String? logoFilename}) {
    final seen = <String>{};
    final candidates = <String>[];

    void add(String path) {
      final normalized = path.toLowerCase().endsWith('.png')
          ? path.replaceAll(RegExp(r'\.png$', caseSensitive: false), '.svg')
          : path;
      if (seen.add(normalized)) candidates.add(normalized);
    }

    if (logoFilename != null && logoFilename.trim().isNotEmpty) {
      add(assetForFilename(logoFilename));
    }

    add(svgForName(brandName));
    add('$_base/default.svg');
    return candidates;
  }

  static bool isSvgAsset(String path) => path.toLowerCase().endsWith('.svg');

  /// True when logo_filename points at a server-managed, user-uploaded logo
  /// (`/assets/brand-logos/brand-{id}.{ext}`) rather than a bundled asset.
  /// Uploaded logos are fetched through the authenticated image proxy.
  static bool isUploadedLogo(String? logoFilename) {
    final trimmed = logoFilename?.trim();
    return trimmed != null && trimmed.startsWith('/assets/brand-logos/brand-');
  }
}
